test/e2e/secret_template_test.go

```diff
@@ -0,0 +1,391 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/ghodss/yaml"
+	sgv1alpha1 "github.com/vmware-tanzu/carvel-secretgen-controller/pkg/apis/secretgen/v1alpha1"
+	sg2v1alpha1 "github.com/vmware-tanzu/carvel-secretgen-controller/pkg/apis/secretgen2/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestSecretTemplate_Full_Lifecycle(t *testing.T) {
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, logger}
+	kubectl := Kubectl{t, env.Namespace, logger}
+
+	testSecretTemplateYaml := `
+---
+apiVersion: secretgen.carvel.dev/v1alpha1
+kind: SecretTemplate
+metadata:
+  name: combined-secret
+spec:
+  inputResources:
+  - name: secret1
+    ref:
+      apiVersion: v1
+      kind: Secret
+      name: secret1
+  - name: secret2
+    ref:
+      apiVersion: v1
+      kind: Secret
+      name: secret2
+  template:
+    type: secret-type
+    data:
+      key1: "$(.secret1.data.key1)"
+      key2: "$(.secret1.data.key2)"
+      key3: "$(.secret2.data.key3)"
+      key4: "$(.secret2.data.key4)"
+`
+
+	testInputResourcesYaml := `
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: secret1
+type: Opaque
+stringData:
+  key1: val1
+  key2: val2
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: secret2
+type: Opaque
+stringData:
+  key3: val3
+  key4: val4
+`
+
+	name := "test-secrettemplate-full-lifecycle"
+	cleanUp := func() {
+		kapp.RunWithOpts([]string{"delete", "-a", name + "-template"}, RunOpts{AllowError: true})
+		kapp.RunWithOpts([]string{"delete", "-a", name + "-inputs"}, RunOpts{AllowError: true})
+	}
+
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("Create Template", func() {
+		kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name + "-template"},
+			RunOpts{StdinReader: strings.NewReader(testSecretTemplateYaml)})
+	})
+
+	logger.Section("Check secret wasn't created and template has ReconcileFailed", func() {
+		out := waitForSecretTemplate(t, kubectl, "combined-secret", sgv1alpha1.Condition{
+			Type:    "ReconcileFailed",
+			Status:  corev1.ConditionTrue,
+			Reason:  "",
+			Message: "cannot fetch input resource secret1: secrets \"secret1\" not found",
+		})
+
+		var secretTemplate sg2v1alpha1.SecretTemplate
+		err := yaml.Unmarshal([]byte(out), &secretTemplate)
+		if err != nil {
+			t.Fatalf("Failed to unmarshal: %s", err)
+		}
+
+		if !reflect.DeepEqual(secretTemplate.Status.Secret.Name, "") {
```

Comment:
please switch to assert.X/require.X (`github.com/stretchr/testify/assert`) -- we've been trying to only use that in our new tests (havent updated all existing tests though).